rock/python.rb

```diff
@@ -0,0 +1,66 @@
+require 'open3'
+
+module Rock
+    # Identify the available python version, per default use the set values in
+    # Autoproj configuration for 'python_executable' and 'python_version',
+    # otherwise checks Autobuild.programs['python'] setting
+    # @param[String] custom_python_bin Path to a custom python version
+    # @returns[[String,String]] path to the python binary and version, throws if
+    #   python cannot be found
+    def self.find_python_bin(custom_python_bin = nil)
+        Autoproj.config.load
+        if Autoproj.config.has_value_for?('python_executable') &&
+                Autoproj.config.has_value_for?('python_version')
+            return Autoproj.config.get('python_executable'),
+                Autoproj.config.get('python_version')
+        end
+
+        finders = [
+            lambda { custom_python_bin },
+            lambda { Autobuild.programs['python'] },
+            lambda { `which python`.strip() }
+        ]
+
+        finders.each do |finder|
+            python_bin = finder.call
+
+            if python_bin && File.exist?(python_bin)
+                cmd = "#{python_bin} -c \"import sys;"\
+                    "version=sys.version_info[:3]; "\
+                    "print('{0}.{1}'.format(*version))\"".strip()
+
+                msg, status = Open3.capture2e(cmd)
+                if status.success?
+                    python_version = msg.strip()
+
+                    if !custom_python_bin
+                        Autoproj.config.set('python_executable', python_bin)
+                        Autoproj.config.set('python_version', python_version)
+                    end
+                    return python_bin, python_version
+                else
+                    Autproj.warn "#{__FILE__} identification of python version"\
+                        " failed: #{msg}"
+                end
+            end
+        end
+        raise RuntimeError("Rock.find_python_bin:"\
+                               "failed to find python")
+    end
+
+    # Allow to update the PYTHONPATH for package, tries to guess the python
+    # binary from Autobuild.programs['python'] and system's default setting
+    # @param[Autobuild::Package] pkg
+    # @param[String] python_bin Path to a custom python version
+    def self.activate_python(pkg, python_bin = nil)
+        begin
+            python_bin, python_version = Rock.find_python_bin(python_bin)
```

Comment:
If I am passing a python bin, I assume that you will use it. You may error if some sanity check fails, but you should *definitely not* override it with something else.

A.k.a. split the "finding the bin", "finding the version" and "checking whether things are rosy" in find_python_bin and handle the multiple code paths either here in activate_python or in a separate `resolve_python` method. This will make the different cases (and how you react to them) a lot clearer. Right now, it's really complex to follow the control flow.

Additionally, one thing you might want to consider is that the python might be installed by Autoproj itself, which means that the existence check would fail on first bootstrap. Not sure what's best (assuming that python is installed prior to using Autoproj seems to be a reasonable workflow)

~~~
def self.resolve_python(ws: Autoproj.workspace, bin: ws.config.get('python_bin', nil), version: ws.config.get('python_version', nil))
    bin ||= find_python_bin
    if !bin
       # Error: can't find
    elsif !File.exist?(bin)
       # Error #{bin} does not exist
    end

    version ||= find_python_version(version)
    unless version
       raise "Error: can't detect version of #{bin}"
    end

    [bin, version]
end

def self.activate_python(pkg, ws: Autoproj.workspace, bin: ws.config.get('python_bin', nil), version: ws.config.get('python_version', nil))
     bin, version = resolve_python(ws: ws, bin: bin, version: version)
     ws.config.set('python_bin', bin)
     ws.config.set('python_version, version)
     ws.env.add_path ...
end
~~~